feat: adding Export Captures button

tranquanghuy0801 opened the pull request:
Resolves #52

tranquanghuy0801 commented:
![Screenshot 2021-09-30 at 22 27 29](https://user-images.githubusercontent.com/29462498/135454788-2d9a99ed-1515-4512-97f9-c3a7feddf63e.png)

nmcharlton commented on src/components/ExportCapture.js:
```diff
@@ -0,0 +1,122 @@
+    setCheckColumns({ ...checkedColumns, [attr]: false });
+  }
+
+  function processDownloadData(captures, selectedColumns) {
```
There's a lot of unnecessary duplication in this function that will make it more difficult to maintain and error prone:

- Use the `renderer` values that are already defined in `columns` rather than defining them all again.
- If you must define new renderers for any reason, make that the only thing you set inside the `if - else` statement. Only include the `formatCapture[attr] = formatCell(...` statement once.

nmcharlton commented on src/components/ExportCapture.js:
```diff
@@ -0,0 +1,122 @@
+        if (attr === 'id' || attr === 'planterId') {
+          formatCapture[attr] = capture[attr];
+        } else if (attr === 'tokenId') {
+          formatCapture[attr] = formatCell(capture, speciesState, attr, (val) =>
```
I'm don't think calling `formatCell` directly here is a good idea. If someone adds another `LinkToWebmap` field, they would need to update this function to avoid breaking CSV export.

Either split out `formatCell` into the content formatting and the presentation, or pass a flag to the function to request a plain text response.

nmcharlton commented on src/components/ExportCapture.js:
```diff
@@ -0,0 +1,122 @@
+            filename={"captures.csv"}
+            asyncOnClick={true}
+            onClick={() => {
+              let filterColumns = Object.entries(checkedColumns).filter((val) => val[1] === true);
```
This can be `const`

nmcharlton commented on src/components/ExportCapture.js:
```diff
@@ -0,0 +1,122 @@
+            className={classes.csvLink}
+            data={downloadData}
+            filename={"captures.csv"}
+            asyncOnClick={true}
```
The `async` functionality isn't working for me (Chrome on MacOS).
A (nearly) empty CSV starts downloading immediately when I click the button.
The next time I click _Download CSV_ I get the data I expect.

nmcharlton commented on src/components/ExportCapture.js:
```diff
@@ -0,0 +1,122 @@
+  const [downloadData, setDownloadData] = useState([]);
+
+  function handleChange(attr) {
+    setCheckColumns({ ...checkedColumns, [attr]: false });
```
This should toggle the value, not always set it to false.

nmcharlton commented on src/components/ExportCapture.js:
```diff
@@ -0,0 +1,122 @@
+              let filterColumns = Object.entries(checkedColumns).filter((val) => val[1] === true);
+              const selectedColumns = Object.fromEntries(filterColumns);
+              capturesContext.getAllCaptures({ filter }).then((response) => {
+                setDownloadData(processDownloadData(response.data, selectedColumns));
```
`processDownloadData` takes a long time and locks up the browser for a few seconds on a fairly powerful machine. On a slower machine this could be a lot longer.

As a minimum, we need to give the user visual feedback that the export is in progress, but even better would be to do the processing in a web worker.
If it's helpful, we can ask the Design team to create a visual workflow.

nmcharlton commented on src/components/ExportCapture.js:
```diff
@@ -0,0 +1,122 @@
+              const selectedColumns = Object.fromEntries(filterColumns);
+              capturesContext.getAllCaptures({ filter }).then((response) => {
+                setDownloadData(processDownloadData(response.data, selectedColumns));
+              });
```
The _Export Captures_ dialog should close when the user clicks `Download CSV`, and be replaced by the visual feedback on progress mentioned in the comment above.

nmcharlton commented on src/components/ExportCapture.js:
```diff
@@ -0,0 +1,122 @@
+      aria-labelledby="form-dialog-title"
+    >
+      <DialogTitle id="form-dialog-title">Export Captures</DialogTitle>
+      <DialogContent>
```
Add a label with of the number of captures to be downloaded. This would be a good place to warn the user that only the first 30,000 will be downloaded.

tranquanghuy0801 commented:
@nmcharlton 
Sorry for the delay on doing this. I was busy on other tasks at work.
For this issue, I have made some changes:

- Remove all conditions for other columns in function `processDownloadData()` except for the `id` and `planterID` as they are applied `<LinkToWebMap>` in formatCell
- Added a linear progress feedback at the end of Export Captures form when you click Download. I just put here for start, I think you might want to change this. You can ask the Design team for this feature and I will update then.
- Fixed the issue of only getting the data at 2nd download time.

nmcharlton commented on src/context/CapturesContext.js:
```diff
@@ -131,6 +132,42 @@ export function CapturesProvider(props) {
+    // if filterInfo contains new values override the defaults in state hooks
+    const {
+      filter = new FilterModel(),
+      orderBy = 'id',
```
We should use the same order as the table: `order: [`${orderBy} ${order}`]`
This is particularly important when the user hits their 20,000 limit.

nmcharlton commented on src/context/CapturesContext.js:
```diff
@@ -131,6 +132,42 @@ export function CapturesProvider(props) {
+    const lbFilter = {
+      where: { ...where },
+      order: [`${orderBy} ${order}`],
+      limit: 20000,
```
We need to warn the user that only the first 20,000 records will be downloaded.

nmcharlton commented:
Well done, this is a lot better!
I think the progress bar in the dialog works well.

One final suggestions: Disable the buttons in the dialog while downloading is in progress and close the dialog when downloading is complete.

nmcharlton reviewed (approved):
Excellent work @tranquanghuy0801, thank you!
